#! /bin/sh -e

# DP: Fix PR java/33765, ignore ZIPEMPTYMAGIC files.

dir=
if [ $# -eq 3 -a "$2" = '-d' ]; then
    pdir="-d $3"
    dir="$3/"
elif [ $# -ne 1 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi
case "$1" in
    -patch)
        patch $pdir -f --no-backup-if-mismatch -p0 < $0
        ;;
    -unpatch)
        patch $pdir -f --no-backup-if-mismatch -R -p0 < $0
        ;;
    *)
        echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
        exit 1
esac
exit 0

2007-11-02  Tom Tromey  <tromey@redhat.com>

	PR java/33765:
	* jcf-parse.c (java_parse_file): Ignore ZIPEMPTYMAGIC files.
	* zipfile.h (ZIPEMPTYMAGIC): New define.

Index: gcc/java/jcf-parse.c
===================================================================
--- gcc/java/jcf-parse.c	(revision 129859)
+++ gcc/java/jcf-parse.c	(revision 129860)
@@ -1946,6 +1946,10 @@
 #endif
 	  parse_zip_file_entries ();
 	}
+      else if (magic == (JCF_u4) ZIPEMPTYMAGIC)
+	{
+	  /* Ignore an empty input jar.  */
+	}
       else
 	{
 	  gcc_unreachable ();
Index: gcc/java/zipfile.h
===================================================================
--- gcc/java/zipfile.h	(revision 129859)
+++ gcc/java/zipfile.h	(revision 129860)
@@ -58,6 +58,7 @@
 #define ZIPDIR_NEXT(ZIPD) \
    ((ZipDirectory*)((char*)(ZIPD)+(ZIPD)->direntry_size))
 #define ZIPMAGIC 0x504b0304	
+#define ZIPEMPTYMAGIC 0x504b0506
 
 extern ZipFile * opendir_in_zip (const char *, int);
 extern int read_zip_archive (ZipFile *);
